[clang][python][test] Move python binding tests to lit framework

rorth opened the pull request:
As discussed in PR #142353, the current testsuite of the `clang` Python bindings has several issues:

- If `libclang.so` cannot be loaded into `python` to run the testsuite, the whole `ninja check-all` aborts.
- The result of running the testsuite isn't report like the `lit`-based tests, rendering them almost invisible.
- The testsuite is disabled in a non-obvious way (`RUN_PYTHON_TESTS`) in `tests/CMakeLists.txt`, which again doesn't show up in the test results.

All these issues can be avoided by integrating the Python bindings tests with `lit`, which is what this patch does:

- The actual test lives in `clang/test/bindings/python/bindings.sh` and is run by `lit`.
- The current `clang/bindings/python/tests` directory (minus the now-superfluous `CMakeLists.txt`) is moved into the same directory.
- The check if `libclang` is loadable (originally from PR #142353) is now handled via a new `lit` feature, `libclang-loadable`.
- The various ways to disable the tests have been turned into `XFAIL`s as appropriate.  This isn't complete and not completely tested yet.
- It keeps the `check-clang-python` target for use by the Clang Python CI.

Tested on `sparc-sun-solaris2.11`, `sparcv9-sun-solaris2.11`, `i386-pc-solaris2.11`, `amd64-pc-solaris2.11`, `i686-pc-linux-gnu`, and `x86_64-pc-linux-gnu`.

rorth committed:
[clang][python][test] Move python binding tests to lit framework

As discussed in PR #142353, the current testsuite of the `clang` Python
bindings has several issues:

- It `libclang.so` cannot be loaded into `python` to run the testsuite, the
  whole `ninja check-all` aborts.
- The result of running the testsuite isn't report like the `lit`-based
  tests, rendering them almost invisible.
- The testsuite is disabled in a non-obvious way (`RUN_PYTHON_TESTS`) in
  `tests/CMakeLists.txt`, which again doesn't show up in the test results.

All these issues can be avoided by integrating the Python bindings tests
with `lit`, which is what this patch does:

- The actual test lives in `clang/test/bindings/python/bindings.sh` and is
  run by `lit`.
- The current `clang/bindings/python/tests` directory (minus the
  now-subperfluous `CMakeLists.txt`) is moved into the same directory.
- The check if `libclang` is loadable (originally from PR #142353) is now
  handled via a new `lit` feature, `libclang-loadable`.
- The various ways to disable the tests have been turned into `XFAIL`s as
  appropriate.  This isn't complete and not completely tested yet.

Tested on `sparc-sun-solaris2.11`, `sparcv9-sun-solaris2.11`,
`i386-pc-solaris2.11`, `amd64-pc-solaris2.11`, `i686-pc-linux-gnu`, and
`x86_64-pc-linux-gnu`.

AaronBallman commented:
CC @llvm/infrastructure-area-team -- I'm not certain who is the best person to review `lit` work; that's sort of the confluence between infraustructure and the rest of the project. Do you have any good suggestions? (We don't have a maintainer listed for it that I could find.)

DeinAlptraum reviewed (commented):
Thank you for the PR, imo this is a much better approach than the previous one.

As I mentioned before, I'm not familiar with the testing infrastructure, but insofar as it works, this looks good to me.

Do note that the libclang CI should also be adapted: `.github/workflows/libclang-python-tests.yml`
The target it uses doesn't exist anymore, and the paths that trigger the workflow should also be changed to remove the old path and include `clang/tests/bindings/python`. You can now also remove the trigger on `clang/CMakeLists.txt`, which should make this run far less often on unrelated PRs.

dwblaikie commented on clang/test/bindings/python/lit.local.cfg:
```diff
@@ -0,0 +1,22 @@
+        conf.lib
+        return True
+    except Exception as e:
+        # Benign error modes.
+        if "wrong ELF class: ELFCLASS32" in str(e):
+            return False
+        elif "No such file or directory" in str(e):
+            return False
+        # Unknown error modes.
+        else:
+            return True
```
This seems backwards - (either the implementation or the comment) if the error modes are benign, wouldn't that mean it's OK/libclang /is/ loadable?

(& also in general, I'd have thought we'd look for specific error modes we can recover from, and anything else means "no recovery" rather than assuming anything other than these bad ones are good?)

rorth replied:
The intent (and it worked for me) is as follows:

- If `libclang.so` cannot be loaded, this can be for two reasons so far:
  - In a 32-bit build on a 64-bit system, the resulting 32-bit `libclang.so` cannot be loaded into the system's 64-bit `python`.  This is as expected, nothing to be done about that, thus the lib is truely considered unloadable.
  - If `libclang.so` is missing completely, this must be because building it has been disabled somehow (probably with `-DLLVM_ENABLE_PIC=OFF`.  Again, this is intentional and nothing to be done.
- However, in other cases there may well be a bug lurking here.  Consider (just for the sake of argument) a `libclang.so` matching `python`'s ELF class, but built with missing dependencies so it's not self-contained and cannot be loaded.  This is simply a bug that could/should be fixed.  Therefore, I return `True` for unknown/unhandled errors, which lets the test to `FAIL`, forcing investigation.  The failure may be benign yet not currently handled, or there may be a real bug that needs fixing.

If I always return `False` on a failure to load `libclang.so`, such bugs would go unnoticed.

dwblaikie replied:
Ah, OK. Some more comments might be helpful - like "if the failure is unexpected, return true, allowing the test to run and fail so developers can see the failure" or something like that?

rorth replied:
I think the update addresses your concern.  I also decided to log unexpected errors from `lit.local.cfg` to avoid every developer encountering this having to add their own debug code.

DeinAlptraum commented:
@rorth have you tested that your `is_libclang_loadable` works as intended w.r.t the "Benign error modes"?
Do correct me if I made a mistake here, but I just tried testing this by manually adding `raise Exception("wrong ELF class: ELFCLASS32")` at the start of `get_cindex_library` and then running the tests via `build/bin/llvm-lit clang/test/bindings/python`, and this just gives me a regular `FAIL`. Adding some debugging statements into `is_libclang_loadable` shows that the `except` block is never entered for some reason...

rorth commented:
> Thank you for the PR, imo this is a much better approach than the previous one.

Thanks.  The previous approach was just a minimal fix (hack?) to avoid breaking the build in a specific situation.  This one (if completed) should handle all known issues with Python bindings testing.

> As I mentioned before, I'm not familiar with the testing infrastructure, but insofar as it works, this looks good to me.

It will certainly need more testing in situations currently covered by `RUN_PYTHON_TESTS` that I known nothing about and probably cannot test myself.

> Do note that the libclang CI should also be adapted: `.github/workflows/libclang-python-tests.yml` The target it uses doesn't exist anymore, and the paths that trigger the workflow should also be changed to remove the old path and include `clang/tests/bindings/python`. You can now also remove the trigger on `clang/CMakeLists.txt`, which should make this run far less often on unrelated PRs.

Ah, I see.  I've made the obvious adjustments locally, but am at a bit of a loss how to handle `build_target` now.  Doing this manually would be an invocation of `bin/llvm-lit clang/test --filter=bindings/python` or some such.  Maybe a new implementation of the `check-clang-python` target needs to be introduced? Don't know if/how I can test this myself.

DeinAlptraum commented:
> I've got no explanation why your explicit raise didn't come through (but then I known very little Python).

Strange, I'll see if I can find out why, but as long as it works also for the expected failure case for you, that's good enough for me.

> am at a bit of a loss how to handle build_target now [...] Don't know if/how I can test this myself.

All the `check-clang-python` did is depend on the `libclang` target and then call unittest, so I think it should be enough to change the `build_target` to `libclang`, and then add another step to the workflow that calls `lit` as you described.
Feel free to abuse the CI to test this until it works.

rorth committed:
- Adjust `libclang-python-tests.yml`.
- Handle `LLVM_USE_SANITIZER` in `lit.local.cfg`.
- Fix SPARC handling, restrict to Linux/sparc*.
- Explain failure modes, log unhandled ones.

rorth commented:
Two issues are worth mentioning about the updated PR:
- Although I'd originally disabled the Clang Python tests on SPARC in [[python, tests] Disable Clang Python tests on SPARC](https://reviews.llvm.org/D60046), they now `PASS` on Solaris/sparcv9, while on Linux/sparc64 `python` `SEGV`s when loading `libclang.so`.  In a `Debug` build, however, I get a `SIGBUS` which seems to point to an FFI issue again, so I'm restricting the `XFAIL` to Linux/sparc64.
- I'm now logging unhandled failures from `lit.local.cfg` so there's a clear indication what's wrong and where to look.

DeinAlptraum commented:
> I wonder if there's a way to recover from my mess, like (in my local repo)

What's preventing you from doing exactly that?

> The weird thing is that build_target isn't even documented in [Workflow syntax for GitHub Actions](https://docs.github.com/en/actions/writing-workflows/workflow-syntax-for-github-actions)

The (original) workflow contains the following:
```
    uses: ./.github/workflows/llvm-project-tests.yml
    with:
      build_target: check-clang-python
      projects: clang
      # There is an issue running on "windows-2019".
      # See https://github.com/llvm/llvm-project/issues/76601#issuecomment-1873049082.
      os_list: '["ubuntu-24.04"]'
      python_version: ${{ matrix.python-version }}
```
which is the definition of a single step in the bulidflow. That step executes the `/.github/workflows/llvm-project-tests.yml` action (an action we've defined ourselves and can be found under that path in the repo). Everything after the `with:` part are the inputs to that action, i.e. the parameters to that function if you will. You won't find `build_target` in any official Github documentation because it's an input to that specific action: if you need any details about its definition or usage, you can take a look at the `llvm-project-tests` action. Happy to answer any questions if anything else is unclear/you need help with any part

DeinAlptraum reviewed (commented):
LGTM now. I'd just like the libclang Python CI to run, but for some reason it doesn't seem to trigger...

DeinAlptraum commented on .github/workflows/libclang-python-tests.yml:
```diff
@@ -33,7 +33,9 @@ jobs:
     uses: ./.github/workflows/llvm-project-tests.yml
     with:
-      build_target: check-clang-python
+      build_target: libclang
+      run: |
+        llvm-lit -v tools/clang/test --filter=bindings.sh
```
Took me a while, the workflow probably doesn't run because there's an error here:
So far the workflow consisted of a single step, so there was no need to open a `steps` section, but since we have two steps this now becomes necessary. I.e. you'd need something like this:
```
    steps:
      - name: Bulid Libclang
        uses: ./.github/workflows/llvm-project-tests.yml
        with:
          build_target: libclang
          projects: clang
          # There is an issue running on "windows-2019".
          # See https://github.com/llvm/llvm-project/issues/76601#issuecomment-1873049082.
          os_list: '["ubuntu-24.04"]'
          python_version: ${{ matrix.python-version }}
      - name: Run tests
        run: llvm-lit -v tools/clang/test --filter=bindings.sh
```

rorth replied:
That's been handled by restoring an updated `check-clang-python` target.

DeinAlptraum reviewed (changes requested):
I think I found the reason why the CI doesn't run. Unfortunately it made me realize that the changes here might require some broader adaptions...

So far the CI was a single-step workflow, that just calls the `llvm-project-tests` workflow with some parameters. But by design that workflow is meant to be run standalone (it is the only step in the pipeline) so it does one-off things, such as determining itself which runner image to use. But once we add a second step in our own workflow, we cannot reasonably delegate the runner image choice to `llvm-project-tests` anymore...

Due to that, I'm not sure if the current approach can be made workable at all without larger changes to our CI setup.
@rorth I know this is a bit much to figure out if you don't have experience with GHA. I will do some experiments to try and figure out if/how this can be solved, but I'm a bit busy atm, so not sure when I'll get around to that. Feel free to abuse the CI to try things out yourself, or to ask if you want any guidance.

rorth committed:
Revert libclang-python-tests.yml changes.
Reintroduce check-clang-python.

DeinAlptraum commented:
I agree, this might be the easier way. Sounds good to me.
Could you resolve the conflict so we can also confirm that the CI runs as usual?

rorth commented:
I'll try.  Hope not to make a total mess again like last time...

DeinAlptraum reviewed (approved):
Great, LGTM!

You can fix the code formatting if you want to, but I can also do this separately... since it's completely unrelated to what you did.

DeinAlptraum commented:
I reverted this in #145774 since it causes build-failures. It seems that the substitution ``CLANG_LIBRARY_PATH=`llvm-config --libdir`` doesn't work in some cases

dyung commented on clang/test/bindings/python/bindings.sh:
```diff
@@ -0,0 +1,35 @@
+
+# Tests fail on Windows, and need someone knowledgeable to fix.
+# It's not clear whether it's a test or a valid binding problem.
+# XFAIL: target={{.*windows.*}}
```
Should this perhaps be `system-windows` instead?

rorth replied:
I suspected something like that initially, but the problem with the `llvm-clang-x86_64-sie-win` buildbot is different: the build logs show
```
LLVM host triple: x86_64-pc-windows-msvc
LLVM default target triple: x86_64-sie-ps5
```
so this is the first instance of a cross-build.  I found that `lit` already has a `native` feature that can be used here, which would resolve `FIXME: Handle CMAKE_CROSSCOMPILING.` in `bindings.sh`: just use
```
XFAIL: !native
```
I've never tried a cross-build of `clang`, so I'd really appreciate if someone with such a setup in place could try this.
